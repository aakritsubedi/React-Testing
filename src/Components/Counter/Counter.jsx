import React, { useState } from 'react';
import Button from '../Button/Button';

export default function Counter() {
  const [counter, setCounter] = useState(0);
  const [errorMessage, setErrorMessage] = useState('');

  const updateCounter = (action) => {
    let factor = 0;
    if (action === 'INCREMENT') {
      factor = 1;
    } else if (action === 'DECREMENT') {
      factor = -1;
    } else {
      setErrorMessage('Unknown action');
    }

    setCounter((prev) => prev + factor);
  };

  return (
    <div className="wrapper">
      <h1>This is counter application</h1>
      <span className="counter">{counter}</span>

      <div className="mt-4">
        <Button
          label="Increment Counter"
          action={() => updateCounter('INCREMENT')}
          className="btn-success"
          id="increment-btn"
        />
        <Button
          label="Decrement Counter"
          action={() => updateCounter('DECREMENT')}
          disabled={counter === 0}
          className="btn-danger"
          id="decrement-btn"
        />
        {errorMessage && <p className="text-danger">{errorMessage}</p>}
      </div>
    </div>
  );
}
