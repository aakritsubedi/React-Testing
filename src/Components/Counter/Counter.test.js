import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Counter from './Counter';

describe('Counter Component Testing', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = mount(<Counter />);
    });

    test('Structure Test', () => {
        expect(wrapper.find('h1').length).toBe(1);
        expect(wrapper.find('button').length).toBe(2);
        expect(wrapper.find('.counter').length).toBe(1);
        expect(wrapper.find('.test-danger').length).toBe(0);
    });

    test('Initial Value Check', function () {
        expect(wrapper.find('.counter').text()).toBe('0');
    });

    test('Counter: renders correctly', () => {
        const tree = renderer.create(<Counter />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('Simulate Initial Click on Decrement Button', () => {
        // Decrement is disabled if counter = 0
        const decrementButton = wrapper.find('#decrement-btn');
        decrementButton.simulate('click');
        expect(wrapper.find('.counter').text()).not.toBe('-1');

        decrementButton.simulate('click');
        expect(wrapper.find('.counter').text()).not.toBe('-2');

        expect(wrapper.find('.counter').text()).toBe('0');
    });

    test('Simulate Click', () => {
        // Increment
        const incrementButton = wrapper.find('#increment-btn');
        incrementButton.simulate('click');
        expect(wrapper.find('.counter').text()).toBe('1');

        incrementButton.simulate('click');
        expect(wrapper.find('.counter').text()).toBe('2');

        // Decrement
        const decrementButton = wrapper.find('#decrement-btn');
        decrementButton.simulate('click');
        expect(wrapper.find('.counter').text()).toBe('1');

        decrementButton.simulate('click');
        expect(wrapper.find('.counter').text()).toBe('0');
    });
});
