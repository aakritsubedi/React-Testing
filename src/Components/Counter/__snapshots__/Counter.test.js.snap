// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component Testing Counter: renders correctly 1`] = `
<div
  className="wrapper"
>
  <h1>
    This is counter application
  </h1>
  <span
    className="counter"
  >
    0
  </span>
  <div
    className="mt-4"
  >
    <button
      className="btn btn-success"
      onClick={[Function]}
      title="Increment Counter"
    >
      Increment Counter
    </button>
    <button
      className="btn btn-disabled"
      disabled={true}
      onClick={[Function]}
      title="Decrement Counter"
    >
      Decrement Counter
    </button>
    
  </div>
</div>
`;
